src/Datagrid/PagerInterface.php

```diff
@@ -60,7 +60,7 @@ public function setQuery($query);
     /**
      * Returns an array of results on the given page.
      *
-     * @return array
+     * @return \Traversable|array
```

Comment:
Why this change ?

You can see the description is saying 
```
Returns an array of results on the given page.
```

Allowing else, even if it's only in the phpdoc should be discuss IMHO.
For instance, this mean we can't use count anymore.

Reply:
Answering this change and the change that is in `src/Datagrid/DatagridInterface.php `

psalm throws such errors
```text
ERROR: InvalidReturnType - src/Datagrid/Datagrid.php:99:21 - The declared return type 'array<array-key, mixed>' for Sonata\AdminBundle\Datagrid\Datagrid::getResults is incorrect, got 'iterable<mixed, mixed>' (see https://psalm.dev/011)
    public function getResults()


ERROR: InvalidReturnStatement - src/Datagrid/Datagrid.php:107:16 - The inferred type 'iterable<mixed, mixed>' does not match the declared return type 'array<array-key, mixed>' for Sonata\AdminBundle\Datagrid\Datagrid::getResults (see https://psalm.dev/128)
        return $this->results;


ERROR: InvalidReturnStatement - src/Datagrid/SimplePager.php:74:20 - The inferred type 'Traversable|non-empty-array<array-key, mixed>' does not match the declared return type 'array<array-key, mixed>' for Sonata\AdminBundle\Datagrid\SimplePager::getResults (see https://psalm.dev/128)
            return $this->results;


ERROR: InvalidReturnStatement - src/Datagrid/SimplePager.php:91:16 - The inferred type '(Doctrine\Common\Collections\ArrayCollection<TKey:Doctrine\Common\Collections\ArrayCollection as array-key, T:Doctrine\Common\Collections\ArrayCollection as mixed>)|Doctrine\Common\Collections\ArrayCollection<array-key, mixed>|mixed' does not match the declared return type 'array<array-key, mixed>' for Sonata\AdminBundle\Datagrid\SimplePager::getResults (see https://psalm.dev/128)
        return $this->results;

```

I thought that the thing I did solves types problem

Reply:
Ok.

`PagerInterface::getResult` have `@return array`
But `Pager::$result` has `@var \Traversable|array|null`.
We're using `\count($this->results);` in the `initializeIterator` so `\Traversable` is not enough.

Then, if we look at `SimplePager`.
```
public function getResults($hydrationMode = null)
    {
        if ($this->results) {
            return $this->results;
        }

        $this->results = $this->getQuery()->execute([], $hydrationMode);
        $this->thresholdCount = \count($this->results);
        if (\count($this->results) > $this->getMaxPerPage()) {
            $this->haveToPaginate = true;

            if ($this->results instanceof ArrayCollection) {
                $this->results = new ArrayCollection($this->results->slice(0, $this->getMaxPerPage()));
            } else {
                $this->results = new ArrayCollection(\array_slice($this->results, 0, $this->getMaxPerPage()));
            }
        } else {
            $this->haveToPaginate = false;
        }

        return $this->results;
    }
```
I see two things:
- `$this->getQuery()->execute([], $hydrationMode);` is currently returning `mixed`. But it relies on doctrine implementations, and I don't think that doctrine return an ArrayCollection.
- if results does not implement ArrayCollection but we have to paginate, we're creating an array collection. It shouldn't.

IMHO the code could be:
```
$this->results = $this->getQuery()->execute([], $hydrationMode);
$this->thresholdCount = \count($this->results);
$this->haveToPaginate = $this->thresholdCount > $this->getMaxPerPage();

if ($this->haveToPaginate) {
    $this->results = \array_slice($this->results, 0, $this->getMaxPerPage()));
}

return $this->results;
```

We could add 
```
if ($this->results instanceof ArrayCollection) {
    $this->results = $this->results->toArray();
}
```
But this shouldn't be necessary to me.

Reply:
@VincentLanglet 
what do you think should be done with this 
https://github.com/sonata-project/SonataAdminBundle/blob/3.x/UPGRADE-3.x.md#deprecated-returning-other-type-than-collection-from-simplepagergetresults

your opinion and this file have "conflicts". What to do then?

Reply:
I found this: https://github.com/sonata-project/SonataAdminBundle/pull/5144/files
So it seems like "my" opinion is the one selected for master.

Reply:
But I don't understand this PR: https://github.com/sonata-project/SonataAdminBundle/pull/5138/files

The changelod is saying
```
### Deprecated
- `SimplePager::getResults` will not return ArrayCollection in next major version (4.0)
```

But the Upgrade not is not saying the same things. There is maybe something to fix ?

Reply:
I don't know. Can you  tag all members of of sonata here?

Reply:
I asked the question in the PR.

But sure, maybe @sonata-project/contributors can help us

Reply:
Maybe looking at master (where we already merged a narrow api about the datagrid directory) can help here.

For the model directory we are working a lot with doctrine collections, but the pager is using mostly arrays. Not sure if that is the correct choice but thats what we merged

Reply:
I read all again and I think the intention was to move to array, but then it was pointed out that using doctrine collections we could have some perf improvements. At the end seems like we left arrays but decided to go with Collections, it is really confusing.

When we decide what is better I can upgrade master branch too.

Reply:
To me looks better with collections BTW, but I dont think thats a change that can be done on 3.x

Reply:
I don't really see a reason to enforce collections.

To me, the choice would be
- Allow only `array`.
- Allow `\Traversable&\Countable&\ArrayAccess|array` to allow Collection and array.

I think that `array` is enough, but I can't argue about the second solution.

Reply:
Okay, hearing all this I will just `baseline` the error, so that there is no bc break and we can choose what to do in v4.x